### ISSUE TITLE
always_declare_return_types should not lint for overridden members

### ISSUE THREAD

**Issue by greglittlefield-wf** [dart-lang/linter#1901]
**Describe the issue**
Overridden members without a return type annotation inherit their return type from the parent class, so the return type is already statically declared.

Redeclaring the return type doesn't aid in static analysis, which is the documented rationale behind this lint:
> Declaring return types for functions helps improve your codebase by allowing the analyzer to more adequately check your code for errors that could occur during runtime.

...so I think it shouldn't apply for this case.

**To Reproduce**
```dart
class A {
  String foo() => 'a';
}

class B extends A {
  @override
  foo() => 'b'; // lint: The method foo should have a return type but doesn't.
}
```

**Expected behavior**
No lint.

**Comment by greglittlefield-wf** [comment]
*Bump* on this issue @pq 

Does this sound like an acceptable change to behavior? If so, I might look into PR-ing it.

**Comment by bwilkerson** [comment]
This is used by both Flutter and pedantic, so we should probably get input from those teams before making changes.

**Comment by eernstg** [comment]
I agree that `always_declare_return_types` should not flag a declaration of an instance member whose return type is obtained by [override inference](https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md#override-inference) (that is, by looking up the overridden declaration(s), checking that they agree, and using the return type from the resulting 'combined' member signature).

As @greglittlefield-wf mentioned, the static analysis is _not_ improved by adding an explicit return type in this case.

Here is an example of a good scenario: A framework _F_ declares a class `C`, an application _A_ uses _F_ and declares one or more subtypes/subclasses of `C`, e.g., `CImpl`. Finally, instances of `CImpl` are passed to code in _F_ and used as a `C`. For instance, _F_ could be Flutter and `C` could be `Widget`.

In this case, `CImpl` and all its member declarations are tightly dependent on _F_, and they are probably rather easy to read because the reader of the code knows the responsibilities of `C` in _F_. So it's a good and safe approach to obtain as many types as possible in the signatures of `CImpl` by override inference, because this is type safe and yet concise, it is consistent with `C`, and it is still easy to read.

Of course, we'd still need to write the return type if there is nothing to override or the overridden return types don't agree, but the lint would still flag those cases. Finally, the return type must be written if we need to make it a subtype of the overridden return type(s); that's not lintable, but we should be OK as long as we don't lint _against_ writing the return type when it's available by override inference.
